File: Dynamic Programming/303. Range Sum Query - Immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
    this.nums = nums;
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {    // Brute force
    return this.nums.slice(left, right + 1).reduce((acc, item) => acc + item, 0);
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */



// In case sumRange gets called thousands time, we can cache the sum from 0 ~ i, so i ~ j = 0 ~ j - 0 ~ i;
var NumArray = function(nums) {
    this.nums = nums;
    this.sums = nums.reduce((acc, item, index) => { 
      acc.push(acc[index] + item);
      return acc;
    }, [0]);
};

NumArray.prototype.sumRange = function(left, right) {
    return this.sums[right + 1] - this.sums[left]; // Note the element is inclusive, so we need to have offset
};

File: Golang/PrefixSum/303.Range-sum-query-immutable.go
package main

type NumArray struct {
	preSum []int
	nums   []int
}

func Constructor(nums []int) NumArray {
	preSum := make([]int, len(nums))
	preSum[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		preSum[i] = nums[i] + preSum[i-1]
	}
	return NumArray{
		preSum: preSum,
		nums:   nums,
	}
}

func (this *NumArray) SumRange(left int, right int) int {
	return this.preSum[right] - this.preSum[left] + this.nums[left]
}
